src/rust/engine/fs/store/src/remote.rs

```diff
@@ -239,71 +230,62 @@ impl ByteStore {
+    let result_future = self.with_byte_stream_client(move |client| {
+      let resource_name = resource_name.clone();
+      let store = store.clone();
+      let f = f.clone();
+      async move {
+        let stream = client
+          .read_opt(
+            &{
+              let mut req = bazel_protos::bytestream::ReadRequest::new();
+              req.set_resource_name(resource_name.clone());
+              req.set_read_offset(0);
+              // 0 means no limit.
+              req.set_read_limit(0);
+              req
+            },
+            call_option(&store.headers, None)?,
+          )
+          .map_err(|err| format!("Error making CAS read request for {:?}: {:?}", digest, err))?;
 
-          let bytes_res = stream
-            .fold(BytesMut::with_capacity(digest.1), move |mut bytes, r| {
-              bytes.extend_from_slice(&r.data);
-              future::ok::<_, grpcio::Error>(bytes)
-            })
-            .compat()
-            .await;
+        let bytes_res = stream
+          .fold(BytesMut::with_capacity(digest.1), move |mut bytes, r| {
+            bytes.extend_from_slice(&r.data);
+            future::ok::<_, grpcio::Error>(bytes)
+          })
+          .compat()
+          .await;
 
-          // We ensure that we hold onto the client until after we've consumed the stream as a
-          // workaround for https://github.com/pingcap/grpc-rs/issues/123
-          std::mem::drop(client);
+        // We ensure that we hold onto the client until after we've consumed the stream as a
+        // workaround for https://github.com/pingcap/grpc-rs/issues/123
+        std::mem::drop(client);
 
-          let maybe_bytes = match bytes_res {
-            Ok(bytes) => Some(bytes.freeze()),
-            Err(grpcio::Error::RpcFailure(grpcio::RpcStatus {
-              status: grpcio::RpcStatusCode::NOT_FOUND,
-              ..
-            })) => None,
-            Err(e) => {
-              return Err(format!(
-                "Error from server in response to CAS read request: {:?}",
-                e
-              ))
-            }
-          };
+        let maybe_bytes = match bytes_res {
+          Ok(bytes) => Some(bytes.freeze()),
+          Err(grpcio::Error::RpcFailure(grpcio::RpcStatus {
+            status: grpcio::RpcStatusCode::NOT_FOUND,
+            ..
+          })) => None,
+          Err(e) => {
+            return Err(format!(
+              "Error from server in response to CAS read request: {:?}",
+              e
+            ))
+          }
+        };
 
-          Ok(maybe_bytes.map(f))
-        }
-      })
-      .await;
+        Ok(maybe_bytes.map(f))
+      }
+    });
 
     if let Some(workunit_state) = workunit_store::get_workunit_state() {
-      workunit_state.store.complete_workunit(span_id);
+      let store = workunit_state.store;
+      with_workunit(store, workunit_name, metadata, result_future, |_, md| md).await
+    } else {
+      result_future.await
     }
```

Comment:
In a followup, it might be useful to expose `with_workunit` as a static method or a macro that would do the equivalent of these 6 lines in one line.